applications/vanilla/controllers/api/DiscussionsApiController.php

```diff
@@ -645,4 +667,68 @@ public function put_bookmark($id, array $body) {
         $result = $this->discussionByID($id);
         return $out->validate($result);
     }
+
+    /**
+     * Set canonical url for a discussion.
+     *
+     * @param int $id The ID of the discussion.
+     * @param array $body The request body.
+     *       Ex: ["canonicalUrl" => "https://mydomain.com/some+path/"]
+     * @throws NotFoundException If unable to find the discussion.
+     * @return array
+     */
+    public function put_canonical($id, array $body) {
+        $this->permission('Garden.SignIn.Allow');
+
+        $this->idParamSchema('in');
+        $in = $this->discussionPutCanonicalSchema('in')->setDescription('Set canonical url for a discussion.');
+        $out = $this->schema($this->discussionSchema(), 'out');
+
+        $body = $in->validate($body);
+
+        $row = $this->discussionByID($id);
+        if (!empty($row['Attributes']['CanonicalUrl'] ?? '')) {
+            throw new \Garden\Web\Exception\ClientException('Canonical url already set for this discussion.', 409);
+        };
+        $categoryID = $row['CategoryID'];
+        if ($row['InsertUserID'] !== $this->getSession()->UserID) {
+            $this->discussionModel->categoryPermission('Vanilla.Discussions.Edit', $categoryID);
+        }
+
+        $attributes = $row['Attributes'];
```

Comment:
If the row's Attributes column was `null` in the database, I believe it'll be `null` here too. The following assignment seems to rely on PHP converting the `null` to an array when referenced it as an array. I think it would be better if we could implement some null-coalescing in this assignment to default it to an empty array.

Reply:
I've updated the code, but it was working fine with nulls.